### PR TITLE
Fix Issue 22861 - build the compiler with PGO

### PR DESCRIPTION
**Pull request by maxhaton**
This introduces a new target to build.d, dmd-pgo, which will build a dmd with PGO
instrumentation, then run either the phobos or dmd testsuite (both are present in the
source code, only the dmd one is set to run at the moment). This data is then merged,
at which point the data is used to build a release + LTO build of dmd.

This resulting build is significantly faster - on the order of 45% on some programs.

ENABLE_PGO=1 can be used to make the build script use the dmd-pgo by default, however
something more subtle may be required to make all dmd builds (i.e. releases)
fully pgo-ified

### CODE REVIEW

File: src/build.d [hunk 1]
```diff
@@ -163,10 +167,13 @@ Command-line parameters
     if (!flags["DFLAGS"].canFind("-color=off") &&
         [env["HOST_DMD_RUN"], "-color=on", "-h"].tryRun().status == 0)
         flags["DFLAGS"] ~= "-color=on";
-
+    string defaultRule()
+    {
+        return usePGO ? "dmd-pgo" : "dmd";
+    }
```
Comment (MoonlightSentinel) [inline src/build.d:173]: This doesn't need to a function. It's only called in line 176

File: src/build.d [hunk 2]
```diff
@@ -475,6 +482,137 @@ alias dmdDefault = makeRule!((builder, rule) => builder
...
+    PGOState pgoState = PGOState(dmdKind);
+
+    alias buildInstrumentedDmd = methodInit!(BuildRule, (rundBuilder, rundRule) => rundBuilder
+        .msg("Build dmd with profile-guided optimization enabled")
+        .deps([dmdExe(null, pgoState.pgoGenerateFlags(), pgoState.pgoGenerateFlags()), dmdConf]));
+    alias testRunner = methodInit!(BuildRule, (rundBuilder, rundRule) => rundBuilder
+        .msg("(DC) RUN.D")
+        .sources([ testDir.buildPath( "run.d") ])
+        .target(env["GENERATED"].buildPath("run".exeName))
+        .command([ env["HOST_DMD_RUN"], "-of=" ~ rundRule.target, "-i", "-I" ~ testDir] ~ rundRule.sources));
```
Comment (MoonlightSentinel) [inline src/build.d:550]: Extract the rule from `runTests` and reuse it here instead of copying it.

File: src/build.d [hunk 3]
```diff
@@ -475,6 +482,137 @@ alias dmdDefault = makeRule!((builder, rule) => builder
...
+
+    alias genData = methodInit!(BuildRule, (rundBuilder, rundRule) => rundBuilder
+        .msg("PGO data generated")
+        .sources([ testDir.buildPath( "run.d") ])
```
Comment (MoonlightSentinel) [inline src/build.d:554]: Neither this nor the following rules depend on `run.d`. The dependency on `testRunner` ensures that this rule is rerun if `run.d` is changed.

File: src/build.d [hunk 4]
```diff
@@ -475,6 +482,137 @@ alias dmdDefault = makeRule!((builder, rule) => builder
...
+        .deps([buildInstrumentedDmd, testRunner])
+        .commandFunction({
+            // Run dmd test suite to get data
+            const scope cmd = [ testRunner.targets[0], "compilable", "-j" ~ jobs.to!string ];
+            log("%-(%s %)", cmd);
+            if (spawnProcess(cmd, null, Config.init, testDir).wait())
+                stderr.writeln("dmd tests failed! This will not end the PGO build because some data may have been gathered");
```
Comment (MoonlightSentinel) [inline src/build.d:561]: Test failures are allowed because of missing tools?

File: src/build.d [hunk 5]
```diff
@@ -475,6 +482,137 @@ alias dmdDefault = makeRule!((builder, rule) => builder
...
+        .target(env["GENERATED"].buildPath("run".exeName))
+        .command([ env["HOST_DMD_RUN"], "-of=" ~ rundRule.target, "-i", "-I" ~ testDir] ~ rundRule.sources));
+
+    alias genData = methodInit!(BuildRule, (rundBuilder, rundRule) => rundBuilder
+        .msg("PGO data generated")
+        .sources([ testDir.buildPath( "run.d") ])
+        .deps([buildInstrumentedDmd, testRunner])
+        .commandFunction({
+            // Run dmd test suite to get data
+            const scope cmd = [ testRunner.targets[0], "compilable", "-j" ~ jobs.to!string ];
```
Comment (MoonlightSentinel) [inline src/build.d:558]: Why restrict this to `compilable` tests?

File: src/build.d [hunk 6]
```diff
@@ -475,6 +482,137 @@ alias dmdDefault = makeRule!((builder, rule) => builder
...
+            //generated/linux/release/64/unittest/test_runner builds the unittests without running them.
+            const scope cmd = ["make", "-C", "../phobos", "-j" ~ jobs.to!string, "-fposix.mak", "generated/linux/release/64/unittest/test_runner", "DMD_DIR="~dmdRepo];
+            log("%-(%s %)", cmd);
+            if (spawnProcess(cmd, null, Config.init).wait())
+                stderr.writeln("Phobos Tests failed! This will not end the PGO build because some data may have been gathered");
+        }));
+    alias finalDataMerge = methodInit!(BuildRule, (rundBuilder, rundRule) => rundBuilder
+        .msg("PGO data has been merged")
+        .sources([ testDir.buildPath( "run.d") ])
+        .deps([/* genData */ genData])
```
Comment (MoonlightSentinel) [inline src/build.d:579]:
```suggestion
        .deps([ genData, genPhobosData ])
```

File: src/build.d [hunk 7]
```diff
@@ -475,6 +482,137 @@ alias dmdDefault = makeRule!((builder, rule) => builder
...
+        .msg("Build dmd with profile-guided optimization enabled")
+        .deps([dmdExe(null, pgoState.pgoGenerateFlags(), pgoState.pgoGenerateFlags()), dmdConf]));
+    alias testRunner = methodInit!(BuildRule, (rundBuilder, rundRule) => rundBuilder
+        .msg("(DC) RUN.D")
+        .sources([ testDir.buildPath( "run.d") ])
+        .target(env["GENERATED"].buildPath("run".exeName))
+        .command([ env["HOST_DMD_RUN"], "-of=" ~ rundRule.target, "-i", "-I" ~ testDir] ~ rundRule.sources));
+
+    alias genData = methodInit!(BuildRule, (rundBuilder, rundRule) => rundBuilder
+        .msg("PGO data generated")
```
Comment (MoonlightSentinel) [inline src/build.d:553]: These messages are printed *before* the `commandFunction` is executed

File: src/build.d [hunk 8]
```diff
@@ -475,6 +482,137 @@ alias dmdDefault = makeRule!((builder, rule) => builder
...
+            if (spawnProcess(cmd, null, Config.init).wait())
+                stderr.writeln("Phobos Tests failed! This will not end the PGO build because some data may have been gathered");
+        }));
+    alias finalDataMerge = methodInit!(BuildRule, (rundBuilder, rundRule) => rundBuilder
+        .msg("PGO data has been merged")
+        .sources([ testDir.buildPath( "run.d") ])
+        .deps([/* genData */ genData])
+        .commandFunction({
+            // Run dmd test suite to get data
+            scope cmd = ["ldc-profdata", "merge", "--output=merged.data"];
```
Comment (MoonlightSentinel) [inline src/build.d:582]: `merged.data` should be emitted to `generated`

File: src/build.d [hunk 9]
```diff
@@ -475,6 +482,137 @@ alias dmdDefault = makeRule!((builder, rule) => builder
...
+            "DFLAGS="~joiner(addArgs, " ").to!string, "--force", "-j"~jobs.to!string];
+            log("%-(%s %)", cmd);
+            if (spawnProcess(cmd, null, Config.init).wait())
+                abortBuild("PGO Compilation failed");
+        }));
+    builder
+        .name("dmd-pgo")
+        .condition(() => PGOState.checkPGO(dmdKind))
+        .msg("Build dmd with merged PGO data")
+        .deps([buildDmdWithData]);
```
Comment (MoonlightSentinel) [inline src/build.d:613]: Merge with the rule defined above?

File: src/build.d [hunk 10]
```diff
@@ -475,6 +482,137 @@ alias dmdDefault = makeRule!((builder, rule) => builder
...
+        .msg("PGO data from phobos generated")
+        .sources([ testDir.buildPath( "run.d") ])
+        .deps([buildInstrumentedDmd])
+        .commandFunction({
+            // Run phobos unittests
+            //TODO makefiles
+            //generated/linux/release/64/unittest/test_runner builds the unittests without running them.
+            const scope cmd = ["make", "-C", "../phobos", "-j" ~ jobs.to!string, "-fposix.mak", "generated/linux/release/64/unittest/test_runner", "DMD_DIR="~dmdRepo];
+            log("%-(%s %)", cmd);
+            if (spawnProcess(cmd, null, Config.init).wait())
```
Comment (MoonlightSentinel) [inline src/build.d:573]: Specify the working directory when using relative paths. Ditto for the other rules.

**Review by MoonlightSentinel — CHANGES_REQUESTED** [review]

File: src/build.d (hunk 4 above)
Comment (maxhaton) [reply src/build.d:561]: Which missing tools? If something has gone catastrophically wrong then I guess the build should stop but frankly I view that as unlikely versus most tests passing but some failing due to CI flakyness

File: src/build.d (hunk 5 above)
Comment (maxhaton) [reply src/build.d:558]: So it runs in less than 20 years. Why would you run the tests?

File: src/build.d (hunk 6 above)
Comment (maxhaton) [reply src/build.d:579]: Not yet

File: src/build.d (hunk 7 above)
Comment (maxhaton) [reply src/build.d:553]: I figured that out at the end, something with the way the script was originally written prior to the commit you see now was making them appear after running

File: src/build.d (hunk 8 above)
Comment (maxhaton) [reply src/build.d:582]: It is

File: src/build.d (hunk 10 above)
Comment (maxhaton) [reply src/build.d:573]: Why?

Comment (maxhaton) [reply src/build.d:573]: The build script maintains an absolute path, why not use it?

File: src/build.d (hunk 9 above)
Comment (maxhaton) [reply src/build.d:613]: No

Comment (maxhaton) [reply src/build.d]: I ended up merging them

**Review by WalterBright — APPROVED** [review]

File: src/build.d [hunk 11]
```diff
@@ -96,6 +98,7 @@ Examples
     ./build.d unittest      # runs internal unittests
     ./build.d clean         # remove all generated files
     ./build.d generated/linux/release/64/dmd.conf
+    ./build.d dmd-pgo       # builds dmd with PGO data, currently only LDC is supported
```
Comment (nordlow) [inline src/build.d:101]: Why isn't this called `release-pgo` or `pgo-release`?

File: src/build.d (hunk 4 above)
Comment (MoonlightSentinel) [reply src/build.d:564]: > Which missing tools? 

Whatever tools are used by postscripts, ... (bash, gdb, ...)

> If something has gone catastrophically wrong then I guess the build should stop but frankly I view that as unlikely versus most tests passing but some failing due to CI flakyness

DMD's test usually aren't flaky (contrary to druntime / phobos).

File: src/build.d (hunk 5 above)
Comment (MoonlightSentinel) [reply src/build.d:561]: More balanced coverage data, especially for the backend (quite a few compilable tests use `-o-`). Running `runnable compilable` (and maybe `fail_compilation`) is fast when omitting the useless permutations for `-O -inline ...` (disabled by passing `ARGS=` to `run.d`)

**Comment by maxhaton** [comment]
https://app.circleci.com/pipelines/github/maxhaton/dmd/60/workflows/7c88436d-89af-4234-9636-fb83b81d8945/jobs/61/parallel-runs/0/steps/0-104 is weird

**Comment by maxhaton** [comment]
@dkorpel @RazvanN7 The CircleCI failure seems to be due to CircleCI rather than this PR (it fails with or without any actual changes in the branch)

File: src/build.d (hunk 1 above)
Comment (maxhaton) [reply src/build.d]: now gone entirely

**Comment by ibuclaw** [comment]
CircleCI passes #13900 :shrug:

**Comment by thewilsonator** [comment]
Superseded by #13900

**Pull request closed without merging**